README.md
# TD output plugin for Embulk

[Treasure Data Service](https://www.treasuredata.com/) output plugin for [Embulk](https://github.com/embulk/embulk)

**NOTICE**:
  * embulk-output-td v0.8.0+ only supports **Embulk v0.9.24**. Embulk v0.9.23 does not work.
  * embulk-output-td v0.5.0+ requires Java 1.8 or higher.
  * embulk-output-td v0.4.0+ only supports **Embulk v0.8.22+**.

## Overview

* **Plugin type**: output
* **Load all or nothing**: yes
* **Resume supported**: no

## Configuration

- **apikey**: apikey (string, required)
- **endpoint**: hostname (string, default='api.treasuredata.com')
- **http_proxy**: http proxy configuration (tuple of host, port, useSsl, user, and password. default is null)
- **use_ssl**: the flag (boolean, default=true)
- **auto_create_table**: the flag for creating the database and/or the table if they don't exist (boolean, default=true)
- **mode**: 'append', 'replace' and 'truncate' (string, default='append')
- **database**: database name (string, required)
- **table**: table name (string, required)
- **session**: bulk_import session name (string, optional)
- **pool_name**: bulk_import session pool name (string, optional)
- **time_column**: user-defined time column (string, optional)
- **unix_timestamp_unit**: if type of "time" or **time_column** is long, it's considered unix timestamp. This option specify its unit in sec, milli, micro or nano (enum, default: `sec`)
- **tmpdir**: temporal directory (string, optional) if set to null, plugin will use directory that could get from System.property
- **upload_concurrency**: upload concurrency (int, default=2). max concurrency is 8.
- **file_split_size**: split size (long, default=16384 (16MB)).
- **stop_on_invalid_record**: stop bulk load transaction if a file includes invalid record (such as invalid timestamp) (boolean, default=false).
- **displayed_error_records_count_limit**: limit the count of the shown error records skipped by the perform job (int, default=10).
- **default_timestamp_type_convert_to**: configure output type of timestamp columns. Available options are "sec" (convert timestamp to UNIX timestamp in seconds) and "string" (convert timestamp to string). (string, default: `"string"`)
- **default_timezone**: default timezone (string, default='UTC')
- **default_timestamp_format**: default timestamp format (string, default=`%Y-%m-%d %H:%M:%S.%6N`)
- **column_options**: advanced: a key-value pairs where key is a column name and value is options for the column.
  - **type**: The type of column when this plugin adds a new column to a TD's table (e.g. `array<string>`). Available options are: `int`, `long`, `float`, `double`, `string`, `array<int>`, `array<long>`, `array<double>`, `array<string>`, `array<array<int>>`. More information can be found: https://tddocs.atlassian.net/wiki/spaces/PD/pages/1083743/Schema+Management. (string, optional)
  - **value_type**: This plugin converts Embulk input data type to msgpack data type that is uploaded to TD. This option controls the msgpack data type which Embulk data in the column is converted to. Available options are: `boolean`, `long`, `double`, `string`, `timestamp`, `array`, `map`. (string, optional)
  - **timezone**: If input column type (embulk type) is timestamp, this plugin needs to format the timestamp value into a SQL string. In this cases, this timezone option is used to control the timezone. (string, value of default_timezone option is used by default)
  - **format**: If input column type (embulk type) is timestamp, this plugin needs to format the timestamp value into a string. This timestamp_format option is used to control the format of the timestamp. (string, value of default_timestamp_format option is used by default)
- **retry_limit**: indicates how many retries are allowed (int, default: 20)
- **retry_initial_interval_millis**: the initial intervals (int, default: 1000)
- **retry_max_interval_millis**: the maximum intervals. The interval doubles every retry until retry_max_interval_millis is reached. (int, default: 90000)
- **additional_http_headers**: add additional headers to the requests (a key & value map, default: null)
- **port**: set port for Http requests. By default will connect to port 443 or 80 if `use_ssl: false` (int, optional)
- **ignore_alternative_time_if_time_exists**: ignore `time_column` and `time_value` in the configuration if a `time` column exists in the input schema. (boolean, default: false)

## Modes
* **append**:
  - Uploads data to existing table directly.
* **replace**:
  - Creates new temp table and uploads data to the temp table first.
  - After uploading finished, the table specified as 'table' option is replaced with the temp table.
  - Schema in existing table is not migrated to the replaced table.
* **truncate**:
  - Creates new temp table and uploads data to the temp table first.
  - After uploading finished, the table specified as 'table' option is replaced with the temp table.
  - Schema in existing table is added to the replaced table.

## Example
Here is sample configuration for TD output plugin.
```yaml
out:
  type: td
  apikey: <your apikey>
  endpoint: api.treasuredata.com
  database: my_db
  table: my_table
  time_column: created_at
  auto_create_table: true
  mode: append
```

### Http Proxy Configuration
If you want to add your Http Proxy configuration, you can use `http_proxy` parameter:
```yaml
out:
  type: td
  apikey: <your apikey>
  endpoint: api.treasuredata.com
  http_proxy: {host: localhost, port: 8080, use_ssl: false, user: "proxyuser", password: "PASSWORD"}
  database: my_db
  table: my_table
  time_column: created_at
  auto_create_table: true
  mode: append
```

### Additional Http headers
```yaml
out:
  type: td
  apikey: <your apikey>
  endpoint: api.treasuredata.com
  database: my_db
  table: my_table
  time_column: created_at
  auto_create_table: true
  mode: append
  additional_http_headers:
    Content_Type: 'application/json'
    foo: bar
```

### Column options
```yaml
out:
  type: td
  apikey: <your apikey>
  endpoint: api.treasuredata.com
  database: my_db
  table: my_table
  time_column: created_at
  auto_create_table: true
  mode: append
  column_options:
    col_array:
      type: array<string>
      value_type: array
    col_long:
      type: string
      value_type: long
    col_timestamp:
      type: string
      value_type: timestamp
      timestamp_format: `%Y-%m-%d %H:%M:%S %z`
      timezone: '-0700'
```

## Install

```
$ embulk gem install embulk-output-td
```

## Build

### Build by Gradle
```
$ git clone https://github.com/treasure-data/embulk-output-td.git
$ cd embulk-output-td
$ ./gradlew gem classpath
```

### Run on Embulk
$ bin/embulk run -I embulk-output-td/lib/ config.yml

## Release

### Upload gem to Rubygems.org

```
$ ./gradlew gem     # create .gem file under pkg/ directory
$ ./gradlew gemPush # create and publish .gem file
```

Repo URL: https://rubygems.org/gems/embulk-output-td

### Upload jars to Bintray.com

```
$ ./gradlew bintrayUpload
```

Repo URL: https://bintray.com/embulk-output-td/maven/embulk-output-td


CHANGELOG.md
# Changelog

Look at [Git annotated tags](https://github.com/treasure-data/embulk-output-td/tags).


src/main/java/org/embulk/output/TdOutputPlugin.java
package org.embulk.output;

import java.io.IOException;
import java.util.List;
import javax.validation.constraints.Min;
import javax.validation.constraints.Max;

import com.google.common.base.Optional;
import com.google.common.base.Throwables;
import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;
import com.treasuredata.api.TdApiClient;
import com.treasuredata.api.TdApiClientConfig;
import com.treasuredata.api.TdApiClientConfig.HttpProxyConfig;
import com.treasuredata.api.TdApiConflictException;
import com.treasuredata.api.TdApiNotFoundException;
import com.treasuredata.api.TdApiException;
import com.treasuredata.api.model.TDBulkImportSession;
import com.treasuredata.api.model.TDBulkImportSession.ImportStatus;
import com.treasuredata.api.model.TDDatabase;
import com.treasuredata.api.model.TDTable;
import org.embulk.config.CommitReport;
import org.embulk.config.Config;
import org.embulk.config.ConfigDefault;
import org.embulk.config.ConfigDiff;
import org.embulk.config.ConfigInject;
import org.embulk.config.ConfigSource;
import org.embulk.config.ConfigException;
import org.embulk.config.Task;
import org.embulk.config.TaskSource;
import org.embulk.output.RecordWriter.FieldWriterSet;
import org.embulk.spi.Exec;
import org.embulk.spi.ExecSession;
import org.embulk.spi.OutputPlugin;
import org.embulk.spi.Schema;
import org.embulk.spi.TransactionalPageOutput;
import org.embulk.spi.time.Timestamp;
import org.joda.time.format.DateTimeFormat;
import org.jruby.embed.ScriptingContainer;
import org.slf4j.Logger;

public class TdOutputPlugin
        implements OutputPlugin
{
    public interface PluginTask
            extends Task
    {
        @Config("apikey")
        public String getApiKey();

        @Config("endpoint")
        @ConfigDefault("\"api.treasuredata.com\"")
        public String getEndpoint();

        @Config("use_ssl")
        @ConfigDefault("true")
        public boolean getUseSsl();

        @Config("http_proxy")
        @ConfigDefault("null")
        public Optional<HttpProxyTask> getHttpProxy();

        //  TODO connect_timeout, read_timeout, send_timeout

        @Config("auto_create_table")
        @ConfigDefault("true")
        public boolean getAutoCreateTable();

        @Config("database")
        public String getDatabase();

        @Config("table")
        public String getTable();

        @Config("session")
        @ConfigDefault("null")
        public Optional<String> getSession();

        @Config("time_column")
        @ConfigDefault("null")
        public Optional<String> getTimeColumn();

        @Config("unix_timestamp_unit")
        @ConfigDefault("\"sec\"")
        public UnixTimestampUnit getUnixTimestampUnit();

        @Config("tmpdir")
        @ConfigDefault("\"/tmp\"")
        public String getTempDir();

        @Config("upload_concurrency")
        @ConfigDefault("2")
        @Min(1)
        @Max(8)
        public int getUploadConcurrency();

        @Config("file_split_size")
        @ConfigDefault("16384") // default 16MB (unit: kb)
        public long getFileSplitSize();

        @ConfigInject
        public ScriptingContainer getJRuby();

        public boolean getDoUpload();
        public void setDoUpload(boolean doUpload);

        public String getSessionName();
        public void setSessionName(String session);
    }

    public interface HttpProxyTask
            extends Task
    {
        @Config("host")
        public String getHost();

        @Config("port")
        public int getPort();

        @Config("use_ssl")
        @ConfigDefault("false")
        public boolean getUseSsl();
    }

    public static enum UnixTimestampUnit
    {
        SEC(1),
        MILLI(1000),
        MICRO(1000000),
        NANO(1000000000);

        private final int unit;

        private UnixTimestampUnit(int unit)
        {
            this.unit = unit;
        }

        public int getFractionUnit()
        {
            return unit;
        }

        @JsonCreator
        public static UnixTimestampUnit of(String s)
        {
            switch (s) {
            case "sec": return SEC;
            case "milli": return MILLI;
            case "micro": return MICRO;
            case "nano": return NANO;
            default:
                throw new ConfigException(
                        String.format("Unknown unix_timestamp_unit '%s'. Supported units are sec, milli, micro, and nano"));
            }
        }

        @JsonValue
        @Override
        public String toString()
        {
            return name().toLowerCase();
        }
    }

    private final Logger log;

    public TdOutputPlugin()
    {
        this.log = Exec.getLogger(getClass());
    }

    public ConfigDiff transaction(final ConfigSource config, final Schema schema, int processorCount,
                                  OutputPlugin.Control control)
    {
        final PluginTask task = config.loadConfig(PluginTask.class);

        // generate session name
        task.setSessionName(buildBulkImportSessionName(task, Exec.session()));

        try (TdApiClient client = newTdApiClient(task)) {
            String databaseName = task.getDatabase();
            String tableName = task.getTable();
            if (task.getAutoCreateTable()) {
                createTableIfNotExists(client, databaseName, tableName);
            } else {
                // check if the database and/or table exist or not
                validateTableExists(client, databaseName, tableName);
            }

            // validate FieldWriterSet configuration before transaction is started
            RecordWriter.validateSchema(log, task, schema);

            return doRun(client, task, control);
        }
    }

    public ConfigDiff resume(TaskSource taskSource,
            Schema schema, int processorCount,
            OutputPlugin.Control control) {
        PluginTask task = taskSource.loadTask(PluginTask.class);
        try (TdApiClient client = newTdApiClient(task)) {
            return doRun(client, task, control);
        }
    }

    private ConfigDiff doRun(TdApiClient client, PluginTask task, OutputPlugin.Control control)
    {
        boolean doUpload = startBulkImportSession(client, task.getSessionName(), task.getDatabase(), task.getTable());
        task.setDoUpload(doUpload);
        control.run(task.dump());
        completeBulkImportSession(client, task.getSessionName(), 0);  // TODO perform job priority

        ConfigDiff configDiff = Exec.newConfigDiff();
        configDiff.set("last_session", task.getSessionName());
        return configDiff;
    }

    public void cleanup(TaskSource taskSource,
            Schema schema, int processorCount,
            List<CommitReport> successCommitReports)
    {
        PluginTask task = taskSource.loadTask(PluginTask.class);
        try (TdApiClient client = newTdApiClient(task)) {
            String sessionName = task.getSessionName();
            log.info("Deleting bulk import session '{}'", sessionName);
            client.deleteBulkImportSession(sessionName);
        }
    }

    private TdApiClient newTdApiClient(final PluginTask task)
    {
        Optional<HttpProxyConfig> httpProxyConfig = newHttpProxyConfig(task.getHttpProxy());
        TdApiClientConfig config = new TdApiClientConfig(task.getEndpoint(), task.getUseSsl(), httpProxyConfig);
        TdApiClient client = new TdApiClient(task.getApiKey(), config);
        try {
            client.start();
        } catch (IOException e) {
            throw Throwables.propagate(e);
        }
        return client;
    }

    private Optional<HttpProxyConfig> newHttpProxyConfig(Optional<HttpProxyTask> task)
    {
        Optional<HttpProxyConfig> httpProxyConfig;
        if (task.isPresent()) {
            HttpProxyTask pt = task.get();
            httpProxyConfig = Optional.of(new HttpProxyConfig(pt.getHost(), pt.getPort(), pt.getUseSsl()));
        } else {
            httpProxyConfig = Optional.absent();
        }
        return httpProxyConfig;
    }

    private void createTableIfNotExists(TdApiClient client, String databaseName, String tableName)
    {
        log.debug("Creating table \"{}\".\"{}\" if not exists", databaseName, tableName);
        try {
            client.createTable(databaseName, tableName);
            log.debug("Created table \"{}\".\"{}\"", databaseName, tableName);
        } catch (TdApiNotFoundException e) {
            try {
                client.createDatabase(databaseName);
                log.debug("Created database \"{}\"", databaseName);
            } catch (TdApiConflictException ex) {
                // ignorable error
            }
            try {
                client.createTable(databaseName, tableName);
                log.debug("Created table \"{}\".\"{}\"", databaseName, tableName);
            } catch (TdApiConflictException exe) {
                // ignorable error
            }
        } catch (TdApiConflictException e) {
            // ignorable error
        }
    }

    private void validateTableExists(TdApiClient client, String databaseName, String tableName)
    {
        try {
            for (TDTable table : client.getTables(databaseName)) {
                if (table.getName().equals(tableName)) {
                    return;
                }
            }
            throw new ConfigException(String.format("Table \"%s\".\"%s\" doesn't exist", databaseName, tableName));
        } catch (TdApiNotFoundException ex) {
            throw new ConfigException(String.format("Database \"%s\" doesn't exist", databaseName), ex);
        }
    }

    private String buildBulkImportSessionName(PluginTask task, ExecSession exec)
    {
        if (task.getSession().isPresent()) {
            return task.getSession().get();
        } else {
            Timestamp time = exec.getTransactionTime(); // TODO implement Exec.getTransactionUniqueName()
            return String.format("embulk_%s_%09d",
                    DateTimeFormat.forPattern("yyyyMMdd_HHmmss").withZoneUTC().print(time.getEpochSecond() * 1000),
                    time.getNano());
        }
    }

    // return false if all files are already uploaded
    private boolean startBulkImportSession(TdApiClient client,
            String sessionName, String databaseName, String tableName)
    {
        log.info("Create bulk_import session {}", sessionName);
        TDBulkImportSession session;
        try {
            client.createBulkImportSession(sessionName, databaseName, tableName);
        } catch (TdApiConflictException ex) {
            // ignorable error
        }
        session = client.getBulkImportSession(sessionName);
        // TODO check associated databaseName and tableName

        switch (session.getStatus()) {
        case UPLOADING:
            if (session.getUploadFrozen()) {
                return false;
            }
            return true;
        case PERFORMING:
            return false;
        case READY:
            return false;
        case COMMITTING:
            return false;
        case COMMITTED:
            return false;
        case UNKNOWN:
        default:
            throw new RuntimeException("Unknown bulk import status");
        }
    }

    private void completeBulkImportSession(TdApiClient client, String sessionName, int priority)
    {
        TDBulkImportSession session = client.getBulkImportSession(sessionName);

        switch (session.getStatus()) {
        case UPLOADING:
            if (!session.getUploadFrozen()) {
                // freeze
                try {
                    client.freezeBulkImportSession(sessionName);
                } catch (TdApiConflictException e) {
                    // ignorable error
                }
            }
            // perform
            client.performBulkImportSession(sessionName, priority);

            // pass
        case PERFORMING:
            log.info("Performing bulk import session '{}'", sessionName);
            session = waitForStatusChange(client, sessionName,
                    ImportStatus.PERFORMING, ImportStatus.READY,
                    "perform");
            log.info("    job id: {}", session.getJobId());

            // pass
        case READY:
            // TODO add an option to make the transaction failed if error_records or error_parts is too large
            // commit
            log.info("Committing bulk import session '{}'", sessionName);
            log.info("    valid records: {}", session.getValidRecords());
            log.info("    error records: {}", session.getErrorRecords());
            log.info("    valid parts: {}", session.getValidParts());
            log.info("    error parts: {}", session.getErrorParts());
            client.commitBulkImportSession(sessionName);

            // pass
        case COMMITTING:
            session = waitForStatusChange(client, sessionName,
                    ImportStatus.COMMITTING, ImportStatus.COMMITTED,
                    "commit");

            // pass
        case COMMITTED:
            return;

        case UNKNOWN:
            throw new RuntimeException("Unknown bulk import status");
        }
    }

    private TDBulkImportSession waitForStatusChange(TdApiClient client, String sessionName,
            ImportStatus current, ImportStatus expecting, String operation)
    {
        TDBulkImportSession importSession;
        while (true) {
            importSession = client.getBulkImportSession(sessionName);

            if (importSession.is(expecting)) {
                return importSession;

            } else if (importSession.is(current)) {
                // in progress

            } else {
                throw new RuntimeException(String.format("Failed to %s bulk import session '%s'",
                            operation, sessionName));
            }

            try {
                Thread.sleep(3000);
            } catch (InterruptedException e) {
            }
        }
    }

    @Override
    public TransactionalPageOutput open(TaskSource taskSource, Schema schema, int taskIndex)
    {
        final PluginTask task = taskSource.loadTask(PluginTask.class);

        RecordWriter closeLater = null;
        try {
            FieldWriterSet fieldWriters = new FieldWriterSet(log, task, schema);
            RecordWriter recordWriter = closeLater = new RecordWriter(task, taskIndex, newTdApiClient(task), fieldWriters);
            recordWriter.open(schema);
            closeLater = null;
            return recordWriter;

        } catch (IOException e) {
            throw Throwables.propagate(e);
        } finally {
            if (closeLater != null) {
                closeLater.close();
            }
        }
    }
}
